ports/stm32f4/common-hal/busio/SPI.c

```diff
@@ -23,37 +24,320 @@
+    #ifdef SPI1
+    if(self->handle.Instance==SPI1) {
+        reserved_spi[0] = false;
+        __HAL_RCC_SPI1_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI2
+    if(self->handle.Instance==SPI2) {
+        reserved_spi[1] = false;
+        __HAL_RCC_SPI2_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI3
+    if(self->handle.Instance==SPI3) {
+        reserved_spi[2] = false;
+        __HAL_RCC_SPI3_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI4
+    if(self->handle.Instance==SPI4) {
+        reserved_spi[3] = false;
+        __HAL_RCC_SPI4_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI5
+    if(self->handle.Instance==SPI5) {
+        reserved_spi[4] = false;
+        __HAL_RCC_SPI5_CLK_DISABLE(); 
+    }
+    #endif
+    #ifdef SPI6
+    if(self->handle.Instance==SPI6) {
+        reserved_spi[5] = false;
+        __HAL_RCC_SPI6_CLK_DISABLE(); 
+    }
+    #endif
+    reset_pin_number(self->sck->pin->port,self->sck->pin->number);
+    reset_pin_number(self->mosi->pin->port,self->mosi->pin->number);
+    reset_pin_number(self->miso->pin->port,self->miso->pin->number);
+    self->sck = mp_const_none;
+    self->mosi = mp_const_none;
+    self->miso = mp_const_none;
+}
+
+static uint32_t stm32_baud_to_spi_div(uint32_t baudrate, uint16_t * prescaler) {
+    static const uint32_t baud_map[8][2] = {
+        {2,SPI_BAUDRATEPRESCALER_2},
+        {4,SPI_BAUDRATEPRESCALER_4},
+        {8,SPI_BAUDRATEPRESCALER_8},
+        {16,SPI_BAUDRATEPRESCALER_16},
+        {32,SPI_BAUDRATEPRESCALER_32},
+        {64,SPI_BAUDRATEPRESCALER_64},
+        {128,SPI_BAUDRATEPRESCALER_128},
+        {256,SPI_BAUDRATEPRESCALER_256}
+    };
+    size_t i = 0;
+    uint16_t divisor;
+    do {
+        divisor = baud_map[i][0];
+        if (baudrate >= (HAL_RCC_GetPCLK2Freq()/divisor)) {
+            *prescaler = divisor;
+            return baud_map[i][1];
+        }
+        i++;
+    } while (divisor != 256);
+    //should never get here
+    mp_raise_RuntimeError(translate("SPI Divisor error"));
+    return 0;
 }
 
 bool common_hal_busio_spi_configure(busio_spi_obj_t *self,
-         uint32_t baudrate, uint8_t polarity, uint8_t phase, uint8_t bits) {
-     return true;
+        uint32_t baudrate, uint8_t polarity, uint8_t phase, uint8_t bits) {
+    //This resets the SPI, so check before updating it redundantly
+    if (baudrate == self->baudrate && polarity== self->polarity 
+        && phase == self->phase && bits == self->bits) return true;
+
+    //Deinit SPI
+    HAL_SPI_DeInit(&self->handle);
+
+    if (bits == 8) self->handle.Init.DataSize = SPI_DATASIZE_8BIT;
+    else if (bits == 16) self->handle.Init.DataSize = SPI_DATASIZE_16BIT;
+    else return false;
+
+    if (polarity) self->handle.Init.CLKPolarity = SPI_POLARITY_HIGH;
+    else self->handle.Init.CLKPolarity = SPI_POLARITY_LOW;
+
+    if (phase) self->handle.Init.CLKPhase = SPI_PHASE_2EDGE;
+    else self->handle.Init.CLKPhase = SPI_PHASE_1EDGE;
```

Comment:
Please use {} around single statements. Omitting them makes it harder to modify without breaking things.